segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,177 +133,352 @@
+        return new MetadataTransaction(this, version.incrementAndGet(), keysToLock);
     }
 
     /**
      * Commits given transaction.
      *
      * @param txn       transaction to commit.
      * @param lazyWrite true if data can be written lazily.
-     * @throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
+     *                  throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
      */
     @Override
-    public void commit(MetadataTransaction txn, boolean lazyWrite) throws StorageMetadataException {
-        commit(txn, lazyWrite, false);
+    public CompletableFuture<Void> commit(MetadataTransaction txn, boolean lazyWrite) {
+        return commit(txn, lazyWrite, false);
     }
 
     /**
      * Commits given transaction.
      *
      * @param txn transaction to commit.
-     * @throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
+     *            throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
      */
     @Override
-    public void commit(MetadataTransaction txn) throws StorageMetadataException {
-        commit(txn, false, false);
+    public CompletableFuture<Void> commit(MetadataTransaction txn) {
+        return commit(txn, false, false);
     }
 
     /**
      * Commits given transaction.
      *
      * @param txn       transaction to commit.
      * @param lazyWrite true if data can be written lazily.
-     * @throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
+     *                  throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
      */
     @Override
-    public void commit(MetadataTransaction txn, boolean lazyWrite, boolean skipStoreCheck) throws StorageMetadataException {
+    public CompletableFuture<Void> commit(MetadataTransaction txn, boolean lazyWrite, boolean skipStoreCheck) {
         Preconditions.checkArgument(null != txn);
-        if (fenced.get()) {
-            throw new StorageMetadataWritesFencedOutException("Transaction writer is fenced off.");
-        }
+        final Map<String, TransactionData> txnData = txn.getData();
 
-        Map<String, TransactionData> txnData = txn.getData();
+        final ArrayList<String> modifiedKeys = new ArrayList<>();
+        final ArrayList<TransactionData> modifiedValues = new ArrayList<>();
 
-        ArrayList<String> modifiedKeys = new ArrayList<>();
-        ArrayList<TransactionData> modifiedValues = new ArrayList<>();
+        val retValue = CompletableFuture.runAsync(() -> {
+                    if (fenced.get()) {
+                        throw new CompletionException(new StorageMetadataWritesFencedOutException("Transaction writer is fenced off."));
```

Comment:
Is it possible to add more context to the error? Like tx id...

Reply:
This is a global error and not transaction specific. It means this instance is fenced off and no more metadata modifications are allowed.